framework/metrics-agent.md

```diff
@@ -0,0 +1,247 @@
+- [AddProperty Input](#addproperty-input)
+
+#### AddDate Input
+
+This is the input to the [AddDate](#adddate) behavior.
+
+The add date input MUST include the following:
+
+- A [label](#label)
+- A [date](#date)
+
+The add date input MAY include the following:
+
+- A [transactionId](#transactionid)
+
+#### AddTime Input
+
+This is the input to the [AddTime](#addtime) behavior.
+
+The add time input MUST include the following:
+
+- A [label](#label)
+- A [duration](#duration)
+
+The add time input MAY include the following:
+
+- A [transactionId](#transactionid)
+
+#### AddCount Input
+
+This is the input to the [AddCount](#addcount) behavior.
+
+The add count input MUST include the following:
+
+- A [label](#label)
+- A [count](#count)
+
+The add count input MAY include the following:
+
+- A [transactionId](#transactionid)
+
+#### AddProperty Input
+
+This is the input to the [AddProperty](#addproperty) behavior.
+
+The add property input MUST include the following:
+
+- A [label](#label)
+- A [value](#value)
+
+The add property input MAY include the following:
+
+- A [transactionId](#transactionid)
+
+### Behaviors
+
+The MetricsAgent Interface MUST support the following behaviors:
+
+- [AddDate](#adddate)
+- [AddTime](#addtime)
+- [AddCount](#addcount)
+- [AddProperty](#addproperty)
+
+#### AddDate
+
+Used to record a specific time value with the same metric name.
+
+#### AddTime
+
+Used to aggregate a sum from multiple time values with the same metric name.
+
+#### AddCount
+
+Used to aggregate a sum from multiple count values with the same metric name.
+
+#### AddProperty
+
+Used to add context/metadata in the form of a key-value pair related to a Metrics reference. 
+
+## Proposed Smithy Model
+
+```smithy
+use aws.polymorph#extendable
+
+@extendable
+resource MetricsAgent {
+  operations: [
+    AddDate,
+    AddTime,
+    AddCount,
+    AddProperty
+  ]
+}
+
+// Operations for different metric types
+operation AddDate {
+  input: AddDateInput,
+  output: AddOutput,
+  errors: [MetricsPutError]
+}
+
+operation AddTime {
+  input: AddTimeInput,
+  output: AddOutput,
+  errors: [MetricsPutError]
+}
+
+operation AddCount {
+  input: AddCountInput,
+  output: AddOutput,
+  errors: [MetricsPutError]
+}
+
+operation AddProperty {
+  input: AddPropertyInput,
+  output: AddOutput,
+  errors: [MetricsPutError]
+}
+
+// Input structures for each operation with flattened values
+structure AddDateInput {
+  @required
+  label: String,
+  @required
+  date: Timestamp,
+  transactionId: String
+}
+
+structure AddTimeInput {
+  @required
+  label: String,
+  @required
+  duration: Long,  // Duration in milliseconds
+  transactionId: String
+}
+
+structure AddCountInput {
+  @required
+  label: String,
+  @required
+  count: Long,
+  transactionId: String
+}
+
+structure AddPropertyInput {
+  @required
+  label: String,
+  @required
+  value: String,
+  transactionId: String
+}
+
+// Common output structure
+structure AddOutput {}
```

Comment:
I get why you might optimize this. But is this really the best choice? why not have a output per operation?